lib/cli/commands/orderbook.ts

```diff
@@ -0,0 +1,183 @@
+import colors from 'colors/safe';
+import { SATOSHIS_PER_COIN } from '../utils';
+
+type FormattedOrderbook = {
+  pairId: string,
+  rows: string[][],
+};
+
+type PriceBucket = {
+  price: number,
+  depth: number;
+};
+
+type OrderbookJson = {
+  pairId: string,
+  sell: PriceBucket[],
+  buy: PriceBucket[],
+};
+
+const COLUMNS = [19, 19, 19, 19];
+const COLUMNS_IN_ORDER_SIDE = COLUMNS.length / 2;
+const HEADER = [
+  { content: colors.green('Buy'), colSpan: 2 },
+  { content: colors.red('Sell'), colSpan: 2 },
+];
+const SECONDARY_HEADER = [
+  colors.green('Price'),
+  colors.green('Depth'),
+  colors.red('Price'),
+  colors.red('Depth'),
+];
+
+const addSide = (buckets: PriceBucket[], precision: number): string[] => {
+  const bucket = buckets.pop();
+  if (bucket) {
+    return [
+      bucket.price.toString(),
+      (bucket.depth / SATOSHIS_PER_COIN).toFixed(precision),
+    ];
+  } else {
+    return Array.from(Array(COLUMNS_IN_ORDER_SIDE)).map(() => '');
+  }
+};
+
+export const createOrderbook = (orders: ListOrdersResponse.AsObject, precision: number) => {
+  const formattedOrderbooks: FormattedOrderbook[] = [];
+  orders.ordersMap.forEach((tradingPair) => {
+    const buy = createOrderbookSide(tradingPair[1].buyOrdersList, false, precision);
+    const sell = createOrderbookSide(tradingPair[1].sellOrdersList, true, precision);
+    const totalRows = buy.length < sell.length
+      ? sell.length : buy.length;
+    const orderbookRows = Array.from(Array(totalRows))
+      .map(() => {
+        return addSide(buy, precision).concat(addSide(sell, precision));
+      });
+    formattedOrderbooks.push({
+      pairId: tradingPair[0],
+      rows: orderbookRows,
+    });
+  });
+  return formattedOrderbooks;
+};
+
+const createTable = () => {
+  const table = new Table({
+    colWidths: COLUMNS,
+  }) as HorizontalTable;
+  table.push(HEADER);
+  table.push(SECONDARY_HEADER);
+  return table;
+};
+
+const displayOrderbook = (orderbook: FormattedOrderbook) => {
+  const table = createTable();
+  orderbook.rows.forEach(row => table.push(row));
+  console.log(colors.underline(colors.bold(`\nTrading pair: ${orderbook.pairId}`)));
+  console.log(table.toString());
+};
+
+const displayTables = (orders: ListOrdersResponse.AsObject, argv: Arguments) => {
+  createOrderbook(orders, argv.precision).forEach(displayOrderbook);
+};
+
+const getAvailablePrices = (orders: Order.AsObject[], sellSide: boolean, count = 8): number[] => {
+  const sortedOrders = [
+    ...new Set(
+      orders.map(order => order.price)
+        .sort((a, b) => a - b),
+    ),
+  ];
+  if (sellSide) {
+    // display sell wall in descending order
+    sortedOrders.reverse();
+  }
+  return sortedOrders.splice(0, count);
+};
+
+const getPriceBuckets = (
+  orders: Order.AsObject[],
+  availablePrices: number[],
+  existingBuckets: PriceBucket[] = [],
+): PriceBucket[] => {
+  // go through all the available price buckets
+  const price = availablePrices.shift();
```

Comment:
I think it's slightly more performant to loop through the array without changing it (as is done with `shift`). I think this would mean abandoning the recursion in place of something like `forEach`, adding orders to a given bucket until you reach the next price point, then starting with a new bucket. It would be interesting to confirm that since you've taken benchmarks already. I think calling `filter` in each loop might be a bit expensive performance-wise too.

Reply:
I tried doing it with `.forEach`, but it didn't read/feel as good as the recursive version. I don't think we'll have performance issues unless we start having hundreds of thousands of orders - I'd be OK addressing the UI/CLI layer performance issues once that happens.

Reply:
Alright, not a huge deal. Like you said though it's maybe something we can revisit in the future.